ably/rest_client.go

```diff
@@ -357,9 +357,10 @@ func (c *RestClient) doWithHandle(r *Request, handle func(*http.Response, interf
 				(c.opts.FallbackHostsUseDefault ||
 					strings.HasPrefix(req.URL.Host, defaultOptions.RestHost) ||
 					c.opts.FallbackHosts != nil) {
-				fallback := defaultOptions.FallbackHosts
-				if c.opts.FallbackHosts != nil {
-					fallback = c.opts.FallbackHosts
+				fallback, error := c.opts.getFallbackHosts()
```

Comment:
If the options are invalid, this should be detected earlier, ie. on `NewREST` or `NewRealtime`.

Reply:
It's invalid only in the case we are using fallbackHosts ... there might be the case when we might not need fallbackHosts. But, I will check if it can be added at the top level.

Reply:
As per [RSC15k](https://docs.ably.io/client-lib-development-guide/features/#RSC15k), there will be an instant error when retried on fallback hosts that do not qualify according to the configuration

Reply:
> (RSC15b) The fallback behavior described by this section, RSC15, only applies when either:
> (RSC15b1) ClientOptions#restHost has not been set to an explicit value (see RSC11a) and ClientOptions#port is not set and ClientOptions#tlsPort is not set
> (RSC15b2) An array of ClientOptions#fallbackHosts is provided
> (RSC15b3) the deprecated ClientOptions#fallbackHostsUseDefault option is set to true
> (RSC15k) When none of the statements in RSC15b are true then host fallback does not apply, and failing HTTP requests that would have qualified for a retry against a fallback host will instead result in an error immediately

This is a whole different thing than [TO3k7](https://docs.ably.io/client-lib-development-guide/features/#TO3k7), which says (emphasis mine):

> (TO3k7) fallbackHostsUseDefault boolean (deprecated) – optionally allows the default fallback hosts [a-e].ably-realtime.com to be used when environment is not production or a custom realtime or REST host endpoint is being used. **It is never valid to configure fallbackHost and set fallbackHostsUseDefault to true**. The fallbackHostsUseDefault option is deprecated and future library releases will ignore any supplied value

RSC15k is about the effects of _valid_ fallback hosts options, while TO3k7 is about what makes  invalid.

Now, it is left unspecificed what should happen when invalid options are provided, but given (per [RSC1b](https://docs.ably.io/client-lib-development-guide/features/#RSC1b)) that we already validate _some_ options at instantiation, I think it's worth validating _all_ options at that point.

It's definitely better dev experience to get early, single feedback on misconfiguration, which is a coding mistake that can just be fixed, than to carry on only to have fallback hosts feature don't work when you actually need it.

Reply:
@lmars what do you think about this one?

Reply:
I think the check should be in the constructor yes, I believe that's what we do in other client libs too.